zkevm-circuits/src/tx_circuit.rs

```diff
@@ -0,0 +1,542 @@
+            error!("Invalid 's' signature value");
+            e
+        })?;
+    // msg = rlp([nonce, gasPrice, gas, to, value, data, sig_v, r, s])
+    let mut stream = RlpStream::new_list(9);
+    stream
+        .append(&tx.nonce)
+        .append(&tx.gas_price)
+        .append(&tx.gas)
+        .append(&tx.to.unwrap_or_else(Address::zero))
+        .append(&tx.value)
+        .append(&tx.data.0)
+        .append(&chain_id)
+        .append(&0u32)
+        .append(&0u32);
+    let msg = stream.out();
+    let msg_hash: [u8; 32] = Keccak256::digest(&msg)
+        .as_slice()
+        .to_vec()
+        .try_into()
+        .expect("hash length isn't 32 bytes");
+    let v = (tx.v - 35 - chain_id * 2) as u8;
+    let pk = recover_pk(v, &tx.r, &tx.s, &msg_hash)?;
+    // msg_hash = msg_hash % q
+    let msg_hash = BigUint::from_bytes_be(msg_hash.as_slice());
+    let msg_hash = msg_hash.mod_floor(&*SECP256K1_Q);
+    let msg_hash_le = biguint_to_32bytes_le(msg_hash);
+    let msg_hash = ct_option_ok_or(secp256k1::Fq::from_repr(msg_hash_le), Error::Synthesis)
+        .map_err(|e| {
+            error!("Invalid msg hash value");
+            e
+        })?;
+    Ok(SignData {
+        signature: (sig_r, sig_s),
+        pk,
+        msg_hash,
+    })
+}
+
+// TODO: Deduplicate with
+// `zkevm-circuits/src/evm_circuit/table.rs::TxContextFieldTag`.
+/// Tag used to identify each field in the transaction in a row of the
+/// transaction table.
+#[derive(Clone, Copy, Debug)]
+pub enum TxFieldTag {
+    /// Unused tag
+    Null = 0,
+    /// Nonce
+    Nonce,
+    /// Gas
+    Gas,
+    /// GasPrice
+    GasPrice,
+    /// CallerAddress
+    CallerAddress,
+    /// CalleeAddress
+    CalleeAddress,
+    /// IsCreate
+    IsCreate,
+    /// Value
+    Value,
+    /// CallDataLength
+    CallDataLength,
+    /// TxSignHash: Hash of the transaction without the signature, used for
+    /// signing.
+    TxSignHash,
+    /// CallData
+    CallData,
+}
+
+#[derive(Clone, Debug)]
+struct TxCircuitConfig<F: Field> {
+    tx_id: Column<Advice>,
+    tag: Column<Advice>,
+    index: Column<Advice>,
+    value: Column<Advice>,
+    sign_verify: SignVerifyConfig<F>,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> TxCircuitConfig<F> {
+    fn new(meta: &mut ConstraintSystem<F>) -> Self {
+        let tx_id = meta.advice_column();
+        let tag = meta.advice_column();
+        let index = meta.advice_column();
+        let value = meta.advice_column();
+        meta.enable_equality(value);
+
+        let power_of_randomness = {
+            let columns = [(); sign_verify::POW_RAND_SIZE].map(|_| meta.instance_column());
+            let mut power_of_randomness = None;
+
+            meta.create_gate("power of randomness", |meta| {
+                power_of_randomness =
+                    Some(columns.map(|column| meta.query_instance(column, Rotation::cur())));
+
+                [0.expr()]
```

Comment:
> What are you constraining in this gate??

Nothing really haha!

The reason for this gate is to get the array of expressions from the power of randomness instance column, so that later on we don't need to query columns everywhere, and can pass the power of randomness array expression everywhere.

It's a trick I copied from here https://github.com/appliedzkp/zkevm-circuits/blob/1b1456609cd6ac9eb54575cdb60a098af3bbe0e5/zkevm-circuits/src/evm_circuit.rs#L384-L396

Reply:
Can we doc this trick?
Before reading this comment, I didn't understand how it works.

Reply:
> Can we doc this trick? Before reading this comment, I didn't understand how it works.

Yes!  I've added a comment both in the `tx_circuit` and `evm_circuit` about it :)